File: LeetCode/iter1/python/ValidateBinarySearchTree.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

#限定范围来比较节点的值是否在其中
class Solution(object):
    def validBSTHelper(self,root,min,max):
        if(root == None):
            return True
        
        if(root.val>min and root.val <max):
            return self.validBSTHelper(root.left,min,root.val) and self.validBSTHelper(root.right,root.val,max)
        else:
            return False
    
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if(root == None):
            return True
            
        return self.validBSTHelper(root.left,-10000000000,root.val) and self.validBSTHelper(root.right,root.val,10000000000)

#利用中序遍历后的链表来比较是否递增
class Solution(object):
    def inorderTraverse(self,root,lst):
        if(root == None):
            return lst
        
        lst = self.inorderTraverse(root.left,lst)
        lst.append(root.val)
        lst = self.inorderTraverse(root.right,lst)
        
        return lst
    
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if(root == None):
            return True
        
        treelst = self.inorderTraverse(root,[])
        
        print(treelst)
        
        lstSize = len(treelst)
        item = treelst[0]
        
        result = True
        
        for i in range(1,lstSize):
            treeItem = treelst[i]
            if(item>=treeItem):
                result = False
                break
            else:
                item = treeItem
                
        return result
        
#测试程序
s = Solution()
root = TreeNode(10)
node1 = TreeNode(5)
node2 = TreeNode(15)

node3 = TreeNode(6)
node4 = TreeNode(20)

root.left = node1
root.right = node2

# node2.left = node3
node2.right = node4

print(s.isValidBST(root))
# print(sys.minsize)

File: LeetCode/iter2/C++/ValidateBinarySearchTree.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <vector>
#include <sstream>
#include <unordered_map>
#include <unordered_set>

using namespace std;

struct TreeNode {
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
	bool isValidBST(TreeNode* root) {
		if (root == NULL)
			return true;

		inOrder(root);

		int start = 1;
		int preItem = v[0];
		bool ret = true;
		while (start < v.size())
		{
			if (preItem >= v[start])
			{
				ret = false;
				break;
			}

			preItem = v[start++];
		}

		return ret;
	}
private:
	vector<int> v;

	void inOrder(TreeNode * root)
	{
		if (root != NULL)
		{
			inOrder(root->left);
			v.push_back(root->val);
			inOrder(root->right);
		}
	}
};

int main()
{
	//TreeNode *root = new TreeNode(10);
	//TreeNode *node1 = new TreeNode(5);
	//TreeNode *node2 = new TreeNode(15);
	//TreeNode *node3 = new TreeNode(6);
	//TreeNode *node4 = new TreeNode(20);

	//root->left = node1;
	//root->right = node2;

	//node2->left = node3;
	//node2->right = node4;

	TreeNode *root = new TreeNode(3);
	TreeNode *node1 = new TreeNode(1);
	TreeNode *node2 = new TreeNode(5);
	TreeNode *node3 = new TreeNode(0);
	TreeNode *node4 = new TreeNode(2);
	TreeNode *node5 = new TreeNode(4);
	TreeNode *node6 = new TreeNode(6);

	root->left = node1;
	root->right = node2;

	node1->left = node3;
	node1->right = node4;

	node2->left = node5;
	node2->right = node6;

	Solution s;

	if (s.isValidBST(root))
		cout << "Yes\n";
	else
		cout << "No\n";

	return 0;
}